libr/search/search.c

```diff
@@ -100,69 +100,153 @@ R_API int r_search_begin(RSearch *s) {
 	return true;
 }
 
-// TODO int -> bool
+// Returns 2 if search.maxhits is reached, 0 on error, otherwise 1
 R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
 	RSearchHit* hit;
 	if (s->align && (addr%s->align)) {
 		eprintf ("0x%08"PFMT64x" unaligned\n", addr);
-		return true;
+		return 1;
 	}
 	if (!s->contiguous) {
 		if (kw->last && addr == kw->last) {
 			kw->count--;
 			kw->last = s->bckwrds ? addr : addr + kw->keyword_length;
 			eprintf ("0x%08"PFMT64x" Sequencial hit ignored.\n", addr);
-			return true;
+			return 1;
 		}
 	}
+	// kw->last is used by string search, the right endpoint of last matcch (forward search), to honor search.overlap
 	kw->last = s->bckwrds ? addr : addr + kw->keyword_length;
+	kw->count++;
+	s->nhits++;
 	if (s->callback) {
-		return s->callback (kw, s->user, addr);
+		int ret = s->callback (kw, s->user, addr);
+		// If callback returns 0 or larger than 1, forwards it; otherwise returns 2 if search.maxhits is reached
+		return !ret || ret > 1 ? ret : s->maxhits && s->nhits >= s->maxhits ? 2 : 1;
 	}
 	if (!(hit = r_mem_pool_alloc (s->pool))) {
-		return false;
+		return 0;
 	}
 	hit->kw = kw;
 	hit->addr = addr;
 	r_list_append (s->hits, hit);
-	return true;
+	return s->maxhits && s->nhits >= s->maxhits ? 2 : 1;
 }
 
 // TODO support search across block boundaries
-// Supported search variants: overlap
-R_API int r_search_deltakey_update(void *_s, ut64 from, const ut8 *buf, int len) {
+// Supported search variants: backward, overlap
+R_API int r_search_deltakey_update(RSearch *s, ut64 from, const ut8 *buf, int len) {
 	RListIter *iter;
-	int i, j, count = 0;
-	RSearch *s = (RSearch*)_s;
+	int longest = 0, i, j;
 	RSearchKeyword *kw;
+	RSearchLeftover *left;
+	const int old_nhits = s->nhits;
+	r_list_foreach (s->kws, iter, kw) {
+		longest = R_MAX (longest, kw->keyword_length + 1);
```

Comment:
why kwlen +1, length is not a range

Reply:
For /d , the needle to be searched is one byte longer than the keyword.